bpo-35381 Remove all static state from posixmodule

eduardo-elizondo opened the pull request:
After #9665, this moves the remaining types in posixmodule to be heap-allocated to make it compatible with PEP384 as well as modifying all the type accessors to fully make the type opaque.

The original PR that got messed up a rebase: https://github.com/python/cpython/pull/10854. All the issues in that commit have now been addressed since https://github.com/python/cpython/pull/11661 got committed.

This change also removes any state from the data segment and onto the module state itself.

<!-- issue-number: [bpo-35381](https://bugs.python.org/issue35381) -->
https://bugs.python.org/issue35381
<!-- /issue-number -->


Automerge-Triggered-By: @encukou

eduardo-elizondo committed:
Make Posixmodule use PyType_FromSpec

eduardo-elizondo committed:
Make Posixmodule use PyType_FromSpec

eduardo-elizondo committed:
Added NEWS

eduardo-elizondo committed:
Added NEWS

eduardo-elizondo committed:
Updated hash

eduardo-elizondo committed:
Updated hash

eduardo-elizondo committed:
Addressed PR Issues

eduardo-elizondo committed:
Addressed PR Issues

eduardo-elizondo committed:
Rebased

eduardo-elizondo committed:
Ran argument clinic

eduardo-elizondo committed:
Ran argument clinic

eduardo-elizondo committed:
Added NEWS

eduardo-elizondo committed:
Added NEWS

eduardo-elizondo committed:
Remove INCREF from GenericAlloc

eduardo-elizondo committed:
Merged to master

eduardo-elizondo committed:
Merge branch 'master' into posixmodule-fromspec

eduardo-elizondo committed:
Cleaned up slot acces

eduardo-elizondo committed:
Cleaned up more slot accesses

eduardo-elizondo committed:
Nits

eduardo-elizondo committed:
Fixes

eduardo-elizondo committed:
Run clinic

eduardo-elizondo committed:
Run clinic

encukou commented on Modules/posixmodule.c:
```diff
@@ -6157,33 +6161,38 @@ os_sched_getscheduler_impl(PyObject *module, pid_t pid)
+"sched_param(sched_priority)\n"
+"--\n"
+"\n"
+"Current has only one field: sched_priority\");\n"
```
This has bad grammar and some extra characters. Also, it has only duplicate information; can it be just removed?
```suggestion
```

eduardo-elizondo replied:
I believe this was copy paste from the clinic. Let me update it.

encukou commented on Modules/posixmodule.c:
```diff
@@ -6157,33 +6161,38 @@ os_sched_getscheduler_impl(PyObject *module, pid_t pid)
-    PyObject *res;
-
-    res = PyStructSequence_New(type);
+PyDoc_STRVAR(os_sched_param__doc__,
```
The reason this is changed to manual argument parsing is that it's `#if`-ed,  right?

eduardo-elizondo replied:
Correct

encukou commented on Modules/posixmodule.c:
```diff
@@ -14462,9 +14463,11 @@ INITFUNC(void)
 
         /* initialize scandir types */
-        if (PyType_Ready(&ScandirIteratorType) < 0)
+        ScandirIteratorType = (PyTypeObject *)PyType_FromSpec(&ScandirIteratorType_spec);
```
This assumes the `INITFUNC` is run exactly once per process. The classes will be shared across subinterpreters, losing most benefits of heap types.

eduardo-elizondo replied:
This is now addressed with the module state

encukou committed:
Improve error handling

encukou commented on Modules/posixmodule.c:
```diff
@@ -14428,8 +14423,11 @@ INITFUNC(void)
-        StatResultType->tp_new = statresult_new;
+        /* Add a custom __new__ to the structsequence */
+        structseq_new = (newfunc)PyType_GetSlot((PyTypeObject *)StatResultType, Py_tp_new);
+        dunder_new = PyDescr_NewClassMethod((PyTypeObject *)StatResultType, &StatResultType_dunder_new);
```
`PyDescr_NewClassMethod` can fail; this is missing error handling.

eduardo-elizondo replied:
Thanks! Fixed

encukou commented on Modules/posixmodule.c:
```diff
@@ -14428,8 +14423,11 @@ INITFUNC(void)
+        /* Add a custom __new__ to the structsequence */
+        structseq_new = (newfunc)PyType_GetSlot((PyTypeObject *)StatResultType, Py_tp_new);
+        dunder_new = PyDescr_NewClassMethod((PyTypeObject *)StatResultType, &StatResultType_dunder_new);
+        PyObject_SetAttrString((PyObject *)StatResultType, "__new__", dunder_new);
```
I don't quite understand this. Why is the `__new__` switched?

eduardo-elizondo replied:
This is the current behavior, so my change preserves this. If the question is: why was the code was written like this in the first place, that's a whole other topic!!

The TL;DR is that this wanted to add some extra checks / error handling so it pulled tp_new out and wrapped it around another function

encukou reviewed (commented):
As Jeroen pointed out in the issue, PEP 384 (Stable ABI) isn't very relevant to the standard library. PEP 489 (Multi-phase init) would bring benefits (better subinterpreter support), but this isn't quite there.
Is there a practical benefit to this change?

eduardo-elizondo committed:
Remove static from posixmodule.c

eduardo-elizondo committed:
Merge branch 'posixmodule-fromspec' of https://github.com/eduardo-elizondo/cpython into posixmodule-fromspec

eduardo-elizondo committed:
Address Comments

eduardo-elizondo committed:
Address Comments

eduardo-elizondo committed:
Fix Windows build

encukou committed:
Use interned strings for constants in module state

encukou committed:
Use descriptor directly, rather than look it up by name

encukou committed:
Bring _Py_IDENTIFIER(__fspath__) back

CPython doesn't yet have a good replacement for _Py_IDENTIFIER, but
reimplementing _PyObject_LookupSpecial in each module that needs it
is not the way to go.
This will need to be changed after we figure out _Py_IDENTIFIER more
systematically.

encukou committed:
Use the tp_free slot directly

encukou committed:
Remove duplicate function

encukou committed:
Use tp_new directly

encukou committed:
Don't call PyState_AddModule

I'm proposing a PyState_AddModule doc update to make things clearer:
https://github.com/python/cpython/pull/16101

eduardo-elizondo committed:
Add __new__ changes

eduardo-elizondo committed:
Merge branch 'master' into posixmodule-fromspec

eduardo-elizondo committed:
Regenerate clinic

eduardo-elizondo committed:
Nits

eduardo-elizondo committed:
Merge to master

eduardo-elizondo committed:
Revert tp_new changes

eduardo-elizondo committed:
Rerun argument clinit

eduardo-elizondo committed:
More fixes

encukou committed:
Re-add newlines and fix docstring for os.sched_param.__new__

encukou committed:
Code style nitpicks

See https://www.python.org/dev/peps/pep-0007/

encukou reviewed (approved):
I had a few cosmetic nitpicks, but I don't think they need a review round-trip, so I'll just merge with them.

Thank you very much for this change!

vstinner commented on Modules/posixmodule.c:
```diff
@@ -537,7 +538,7 @@ _Py_Uid_Converter(PyObject *obj, void *p)
         PyErr_Format(PyExc_TypeError,
                      "uid should be integer, not %.200s",
-                     Py_TYPE(obj)->tp_name);
+                     _PyType_Name(Py_TYPE(obj)));
```
tp_name and _PyType_Name() are not the same. There were some discussions between fully qualified names (module.name1.name2.name3) vs short name (name3), but no consensus was found. Well, I should reopen a discussion ;-) I prefer to provide longest names in error messages to ease debug.

eduardo-elizondo replied:
@vstinner I actually debated over that for a bit! I think we are missing something like: `PyType_QualName`. I can do a quick PR for that but I don't know if there will be push back from other devs. Thoughts?

encukou replied:
Reopening the discussion sounds like a good idea.

encukou replied:
Was it reopened? A discussion could be useful in https://bugs.python.org/issue42035.
